cpp/splinepy/proximity/proximity.hpp

```diff
@@ -0,0 +1,342 @@
+      if (!kdtree_planted_) {
+        // hate to be aggresive, but here it is.
+        splinepy::utils::PrintAndThrowError(
+            "to use InitialGuess::Kdtree, please first plant the kdtree!"
+        );
+      }
+
+      // good to go. ask the tree
+      const int vote = 1;
+      int id[vote];
+      double distance[vote];
+      kdtree_->knnSearch(goal,
+                         vote /* closest neighbor */,
+                         &id[0],
+                         &distance[0]);
+
+      using ReturnType = typename SplineType::ParametricCoordinate_;
+
+      return 
+          grid_points_.template IndexToParametricCoordinate<ReturnType>(id[0]);
+    } else {
+      //well, shouldn't reach here, but to fight a warning, here it is
+      splinepy::utils::PrintAndThrowError("Invalid initial guess option!");
+      return typename SplineType::ParametricCoordinate_{};
+    }// if initial_guess == ...
+  }
+
+  /*!
+   * Builds RHS and fills "eyeders" matrix, which is also required in LHS.
+   *
+   * "eyeders" refer to collection of spline derivatives where only one
+   * parametric dimension is set to one and everythin else zero, which results
+   * in "eye"-like vectorized derivative query
+   *
+   * @params[in] guess current parametric coordinate guess
+   * @params[in] difference result of `GussMinusQuery()`
+   * @params[out] eyeders
+   * @params[out] rhs
+   */
+  void FillEyeDersAndRhs(
+      const typename SplineType::ParametricCoordinate_& guess,
+      const DArrayD_& difference,
+      PxDMatrixD_& eyeders,
+      PArrayD_& rhs) const {
+    // btw, RHS is what's internally called as "df_dxi"
+    double df_dxi_i; // temporary variable to hold sum
+    int j;
+    typename SplineType::Derivative_ derivative_query;
+
+    for (int i{}; i < SplineType::kParaDim; ++i) {
+      // eyeder query formulation
+      derivative_query.fill(splinelib::Derivative{0});
+      ++derivative_query[i];
+      // derivative evaluation
+      auto const &derivative = spline_(guess, derivative_query);
+
+      j = 0;
+      df_dxi_i = 0.;
+      for (const auto& d : derivative) {
+        // cast, since d may be a NamedType.
+        const double d_value = static_cast<double>(d);
+        df_dxi_i += difference[j] * d_value;
+        eyeders[i][j] = d_value;
+        ++j;
+      }
+      rhs[i] = -2. * df_dxi_i; // apply minus here already!
+    }
+  }
+
+  /*!
+   * Builds LHS
+   *
+   * @params[in] guess
+   * @params[in] difference
+   * @params[in] eyeders
+   * @params[out] lhs
+   */
+  void FillLhs(
+      const typename SplineType::ParametricCoordinate_& guess,
+      const DArrayD_& difference,
+      const PxDMatrixD_& eyeders,
+      PxPMatrixD_& lhs) const {
+
+    const PxPMatrixD_ eyedersAAt = splinepy::utils::AAt(eyeders);
+
+    typename SplineType::Derivative_ derivative_query;
+    for (int i{}; i < SplineType::kParaDim; ++i) {
+      for (int j{i}; j < SplineType::kParaDim; ++j) {
```

Comment:
I like that you used my suggestion here :D
As a general use of thumb it would be nice if you used unsigned types in loops were indices are required. That makes it more intuitive imo

Reply:
true I would prefer that. I think I did this because SplineLib splines have tparams (para_dim & dim) as int and i'd get warnings with uint. As we discussed, upcoming PR should re-structure splinepy-cpp-splines and maybe we can just save kParaDim as uint and use replace all the index type with size_t